src/main.rs

```diff
@@ -23,56 +23,57 @@ use tempfile;
 mod cmd_exec;
 mod common;
 mod crypto;
+mod error;
 mod hash;
-mod keylime_error;
 mod secure_mount;
 mod tpm;
 
 use common::config_get;
-use common::emsg;
 use common::set_response_content;
-use futures::future;
-use hyper::rt::Future;
-use hyper::service::service_fn;
+use hyper::service::{make_service_fn, service_fn};
 use hyper::{Body, Method, Request, Response, Server, StatusCode};
 use serde_json::Map;
 use std::collections::HashMap;
-use std::error::Error;
 use std::fs::File;
 use std::io::BufReader;
 use std::io::Read;
 use std::path::Path;
 
-type BoxFut = Box<Future<Item = Response<Body>, Error = hyper::Error> + Send>;
+use error::{Error, Result};
 
 static NOTFOUND: &[u8] = b"Not Found";
 
-fn main() {
+#[tokio::main]
+async fn main() -> Result<()> {
     pretty_env_logger::init();
     // Get a context to work with the TPM
-    let mut ctx = tpm::get_tpm2_ctx();
+    let mut ctx = tpm::get_tpm2_ctx()?;
 
     let cloudagent_ip =
-        config_get("/etc/keylime.conf", "cloud_agent", "cloudagent_ip");
+        config_get("/etc/keylime.conf", "cloud_agent", "cloudagent_ip")?;
     let cloudagent_port =
-        config_get("/etc/keylime.conf", "cloud_agent", "cloudagent_port");
+        config_get("/etc/keylime.conf", "cloud_agent", "cloudagent_port")?;
     let endpoint = format!("{}:{}", cloudagent_ip, cloudagent_port);
 
     info!("Starting server...");
 
     let addr = (endpoint).parse().expect("Cannot parse IP & Port");
 
-    let server = Server::bind(&addr)
-        .serve(|| service_fn(response_function))
-        .map_err(|e| error!("server error: {}", e));
+    let service = make_service_fn(|_| async {
+        Ok::<_, Error>(service_fn(response_function))
+    });
+    let server = Server::bind(&addr).serve(service);
 
     info!("Listening on http://{}", addr);
 
     // run server forever
-    hyper::rt::run(server);
+    //hyper::rt::run(server);
+    server.await?;
+
+    Ok(())
 }
 
-fn response_function(req: Request<Body>) -> BoxFut {
+async fn response_function(req: Request<Body>) -> Result<Response<Body>> {
```

Comment:
Is there any reason for changing to asynchronous , we historically only used async on the verifier (which is one to many) where as the agent has never been a performance factor (as it only deals serving with one verifier typically or talking to the TPM)

Reply:
Yes, because that's technically what you always did with Hyper, it was just hidden behind a `Future`. I just rebased to Hyper 0.13 as part of this, so the types worked out.